### PR TITLE
Initial work of creating a Trainer class.

### PR DESCRIPTION
**Pull request by ssm951**
## Refactor the NPI Assignment
By Sam Meng

### Changelog

* NPI Config class to store common config for model, dataset, and training configuration
* NPI Trainer class separating training logic from model/data initialization, saving of metadata, and visualization.
* NPI Training Models class as the one-stop-shop to load models based on the NPIConfig class
* TODO: Separate visualization code from training and generating of loss data.
* TODO: Add logic to create dataset, and unify everything in one function call.
* TODO: Integrate other's changes into what I have here

### CODE REVIEW

File: src/npi/__main__.py [hunk 1]
```diff
@@ -1,31 +1,58 @@
...
+        trainer.train_adversarial_npi(
+            models, num_epochs, train_data, test_data
+        )
+    elif model == "classifier":
         train_classifier(args)
     else:
         print("Only can train classifier or npi.")
 
+
 def test(args):
```
Comment (ssm951) [inline src/npi/__main__.py:41]: @pwertenb Would you be able to add scripting code here running the test_npi? To run this script, cd to the `src` folder, then run `python -m npi test`. You may have to fix what we currently had since I never tested this.

File: src/npi/__main__.py [hunk 2]
```diff
@@ -1,31 +1,58 @@
...
+    if model == "npi":
+        # test_npi(args)
+        pass
+    elif model == "classifier":
         test_classifier(args)
     else:
         print("Only can train classifier or npi.")
 
+
 def construct(args):
```
Comment (ssm951) [inline src/npi/__main__.py:52]: Trevor, if I recall, you were working on construct_data right? You can add code in here to get construct working. Don't worry about the args for now, just hard code the parameters that work into a variable and we can work from there.

**Review by ssm951 — COMMENTED** [review]
I'm thinking we should merge our changes and write code that can run our changes in the `src/npi/__main__.py`. You can look at what I added in that file for reference.

To run that script correctly, cd to the `src` folder, then run `python -m <function_name>`. For example, to run the train NPI code, I would run `python -m train npi`. You may have to fix what we currently had since I never tested on the other commands. Again, take a look at what I added as reference.
